### ISSUE TITLE
Missing first character of <filename>.xoj when annotating a PDF

### ISSUE THREAD

**Issue by corna** [xournalpp/xournalpp#108]
When I annotate a PDF the default filename has the first character missing
For example, if I open "PART 1.pdf" and I annotate it, the default filename is "ART 1.xoj"
If I reload the PDF, the annotations are missing (I can get them again by loading the xoj file), therefore the bug should be in the saving functions
I'm using the latest git version

**Comment by wbrenna** [comment]
Interesting - I've never had it miss a character for the default filename like that. I tried annotating with a pdf called ```PART 1.pdf``` and it worked OK on this side. It might be system specific - what distribution are you on? And are you using the gtk3 branch of xournalpp, master, etc.?

Regarding saving, you opened the PDF, annotated, and then used export to PDF, but the annotations didn't appear? I've had the opposite happen; this is related to bug #60. Maybe if you can provide a PDF I can get to the bottom of this one more quickly.

**Comment by Corna** [comment]
I'm on Debian Jessie, and, since I have libpoppler v0.26.5, I'm using the glib_new branch
I did not export to PDF, I've left the annotations in the .xoj file. If I load the "clean" PDF it should automatically load also <pdfname>.pdf.xoj, but it doesn't (a file called like that does not exist)

**Comment by wbrenna** [comment]
Ah, I misunderstood the problem. When it's unable to load annotations, is that because the xoj file is missing the first character or did you correct the filename from the default, and it still doesn't load?

And to confirm, this happens for all PDFs that you try to annotate, not just that particular one?

**Comment by Corna** [comment]
If I correct the 1st character, everything works, and it happens to every PDF
I've just tested the gtk3 branch and the issue is still present (btw, I'm using GNOME 3.14)
I've also noticed another bug with the gtk3 branch: if you write lines very fast, the curve is divided in segments (even if I disable event compression). This behaviour doesn't appear in the glib_new branch. Do you want me to open another issue report for this?

**Comment by wbrenna** [comment]
I'll attempt to test this on a Jessie system and see if I can replicate. In the meantime if there are any error/warning messages when you start xournalpp from the command line, that would be helpful.

Right now the gtk3 branch is in a bit of limbo regarding drawing lines. I noticed there were a couple of commits that were accidentally undone and I will have to take a look at it there. So regarding the gtk3 drawing bug, feel free to open another issue, but it will hopefully be tackled when gtk3 gets a threading overhaul!

**Comment by Corna** [comment]
With glib_new branch:
MemoryLeak: sum 0 objects.

With gtk3 branch:
Gtk-Message: GtkDialog mapped without a transient parent. This is discouraged.
MemoryLeak: sum 0 objects.

**Comment by Corna** [comment]
Btw, thanks for this great software! Today an university collegue of mine (Windows user with a Surface) asked me where he could download this software as replacement of OneNote :-)
I hope you'll fix the bugs on gtk3, as it integrates very well on GNOME 3

**Comment by Corna** [comment]
I've also noticed that the autosaves and backups also get weird names...

![xournalpp](https://cloud.githubusercontent.com/assets/6459604/6621025/6e64c4bc-c8d4-11e4-8cee-cb4ae00c7891.png)

**Comment by MarPiRK** [comment]
Do you have some Unicode characters in file path? If so it's similar problem to issue #101. I'm working on this problem right now – for me it has finally worked as it should. There is a little more memory used, but I think, that I'll manage it in future commits. You can try out string_new branch from my forked repo.

**Comment by wbrenna** [comment]
Trying this out is a good idea. It's MarPiRK/xournalpp@342feb66ee8232ba2e752e9d97caeb7078dad6e5.

Hopefully I can get some time shortly to help out with gtk3!

**Comment by MarPiRK** [comment]
I realised few bugs after writing this and it's really early version, so don't expect it working as a charm from the start.

**Comment by Corna** [comment]
I don't have unicode characters (as you can see on my screenshot), but your commits fixed my error, thanks!
I'll keep testing

**Comment by wbrenna** [comment]
That's great - @MarPiRK, once you get your branch fixed up to your satisfaction, maybe make a patch and I can test it out and put it in master.

**Comment by MarPiRK** [comment]
Thanks. (Un)fortunately my patch is based on icu library – it means another dependency, but in return we'll have pretty redundant string subsystem. Besides I very like icu and supporting it now may help in future, when full i18n will be (hopefully) implemented.

**Comment by MarPiRK** [comment]
Can you check out if new version works?

**Comment by Corna** [comment]
I've found some build issues:
 * ./configure returns this warning: "AX_CXX_COMPILE_STDCXX_11: command not found"; after putting ax_cxx_compile_stdcxx_11.m4 (taken from the autoconf-archive) in the m4 folder the warning disappear and the configuration script configures correctly c++11
 * ./configure return another warning: "AX_BOOST_ALGORITHM: command not found"; removing the corresponsing line fixes the error and the compilation works fine (btw, I didn't find any reference to AX_BOOST_ALGORITHM in the autoconf archive)
 * ./configure fails with a "syntax error near [...] AX_BOOST_BASE"; putting ax_boost_base.m4, ax_boost_locale.m4, ax_boost_filesystem.m4 and ax_boost_system.m4 (from the autoconf archive) in the m4 folder fixes the error
 * CSTR must be converted back to c_str to get the python module work
 
Instead of manually putting the m4 files in m4/ the errors can be also fixed by installing autoconf-archive (Debian). In this way, all the required m4 files will be put in /usr/share/aclocal/ and autoreconf can automatically find them. Should we include the m4 scripts in the m4 folder or use the distro's ones?

 Everything else works fine, including a ICU bug that set an incorrect associated PDF file, resulting in a "Missing background" error each time a xoj file was created with the ICU version. Also the backup and autosave filenames are now correct (for example, now it saves an .FILENAME.pdf.autosave.xoj instead of F.ILENAME.pdf.autosave.xoj). Moreover, the missing first character bug is now fixed.
 
Unexpectly, it crashed once with this error log: 
 
 ```
 Date: Sat Mar 21 23:56:09 2015
Error: signal 11
[bt]: (0) xournalpp() [0x64e4d8]
[bt]: (1) /lib/x86_64-linux-gnu/libc.so.6(+0x35180) [0x7f6bd0b79180]
[bt]: (2) /lib/x86_64-linux-gnu/libc.so.6(cfree+0x14) [0x7f6bd0bc0664]
[bt]: (3) xournalpp(_ZN6PDFDocD2Ev+0x116) [0x5cb796]
[bt]: (4) xournalpp(_ZN18XojPopplerDocumentaSERS_+0x2d2) [0x55e902]
[bt]: (5) xournalpp(_ZN8DocumentaSERS_+0xa7) [0x551a67]
[bt]: (6) xournalpp(_ZN7Control8openFileESsi+0x5f8) [0x4d88d8]
[bt]: (7) xournalpp(_ZN7Control15actionPerformedE10ActionType11ActionGroupP9_GdkEventP12_GtkMenuItemP14_GtkToolButtonb+0xbd6) [0x4da3d6]
[bt]: (8) xournalpp(_ZN12AbstractItem9activatedEP9_GdkEventP12_GtkMenuItemP14_GtkToolButton+0xfb) [0x51c45b]
[bt]: (9) /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0(g_closure_invoke+0x145) [0x7f6bd3b86245]
[bt]: (10) /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0(+0x21f6c) [0x7f6bd3b97f6c]
[bt]: (11) /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0(g_signal_emit_valist+0xfd8) [0x7f6bd3ba0778]
[bt]: (12) /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0(g_signal_emit+0x8f) [0x7f6bd3ba09df]
[bt]: (13) /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0(gtk_widget_activate+0x76) [0x7f6bd5898af6]
[bt]: (14) /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0(gtk_menu_shell_activate_item+0xfd) [0x7f6bd579494d]
[bt]: (15) /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0(+0x144ceb) [0x7f6bd5794ceb]
[bt]: (16) /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0(+0x132a3f) [0x7f6bd5782a3f]
[bt]: (17) /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0(g_closure_invoke+0x145) [0x7f6bd3b86245]
[bt]: (18) /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0(+0x21e62) [0x7f6bd3b97e62]
[bt]: (19) /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0(g_signal_emit_valist+0xae5) [0x7f6bd3ba0285]
[bt]: (20) /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0(g_signal_emit+0x8f) [0x7f6bd3ba09df]
[bt]: (21) /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0(+0x249d6c) [0x7f6bd5899d6c]
[bt]: (22) /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0(gtk_propagate_event+0xc4) [0x7f6bd5781184]
[bt]: (23) /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0(gtk_main_do_event+0x3ab) [0x7f6bd578161b]
[bt]: (24) /usr/lib/x86_64-linux-gnu/libgdk-x11-2.0.so.0(+0x5ab5c) [0x7f6bd53f4b5c]
[bt]: (25) /lib/x86_64-linux-gnu/libglib-2.0.so.0(g_main_context_dispatch+0x24d) [0x7f6bd38b0c5d]
[bt]: (26) /lib/x86_64-linux-gnu/libglib-2.0.so.0(+0x49f48) [0x7f6bd38b0f48]
[bt]: (27) /lib/x86_64-linux-gnu/libglib-2.0.so.0(g_main_loop_run+0xc2) [0x7f6bd38b1272]
[bt]: (28) /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0(gtk_main+0xb7) [0x7f6bd5780597]
[bt]: (29) xournalpp(_ZN11XournalMain3runEiPPc+0x6a6) [0x4f1e96]
[bt]: (30) xournalpp(main+0x67) [0x49cee7]
[bt]: (31) /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf5) [0x7f6bd0b65b45]
[bt]: (32) xournalpp() [0x49d317]


Try to get a better stracktrace...
```

**Comment by MarPiRK** [comment]
I think that we should add it for now to repo. These files aren't updated too frequently, so it's no harm.
Technically it wasn't ICU's bug, but I replaced old terrible String class with ICU's UnicodeString and it worked as it should. So the order of changes is: TerribleString -> icu::UnicodeString -> std::string (with boost::locale support).
Do you remember when it crashed?

**Comment by Corna** [comment]
It crashed with std::string + boost::locale, but only once
I'll keep testing

**Comment by MarPiRK** [comment]
I have reproduced this crash on my build – I opened quickly different files (xoj and pdf) and evetually it crashes while loading PDF. I'm working now on this bug.

**Comment by MarPiRK** [comment]
I have commited some changes for further tesing.

**Comment by MarPiRK** [comment]
If you noticed any bugs in string_new, please ensure you have latest version compiled and post issue in my repo.

**Comment by wbrenna** [comment]
Since it looks like PDF export doesn't yet work properly I'll hold off of merging this in until that is resolved, yes?

**Comment by MarPiRK** [comment]
Yes, but I'd like to be really sure about my code before requesting pull and hence I've made really significant changes I have to test pretty much everything. I hope that next week it'll be ready to merge, but I think that on the beginning we should place it in string_new branch, next to stable glib_new master for a while,  so all the people using Xournal++ would test it properly. When we'll be sure it's stable enough, we'd make it master. Then I'll start merging it with gtk3 branch.

**Comment by MarPiRK** [comment]
I think that my string_new branch is ready to be present in main xournalpp repo.

**Comment by MarPiRK** [comment]
@wbrenna Can you create string_new branch, so I can request pull?

**Comment by andreasb242** [comment]
I looked at your changes, my findings:
---
src/control/xml/TextAttribute.cpp:
replace_pair('\"', "&quot;"),
Looks wrong for me, why did backslash the double quote in the single quote?
---
Editing std::string in loops is very inefficient, it should be replaced by stringstream
(I now, this was not better before...)

But in general it looks all OK.

**Comment by andreasb242** [comment]
@MarPiRK 
I saw, you changed the license to GPLv3.

Please don't do that, GPLv3 has some Problems, not compatible with GPLv2, requires to share keys (e.g. if we do an Android Port) which makes Problems.

We should write __GPLv2 or later__.

An example:
http://www.phoronix.com/scan.php?page=news_item&px=MTI4Mjc

**Comment by MarPiRK** [comment]
Uh... I cannot find this kind of replace_pair anywhere...
In my latest commit (MarPiRK/xournalpp@273dcb8b50eb5351dda4e32384a6f5de6e0c26aa) I've made changes to replace_all_chars method, so it uses stringstream. Do you see any other places, which can be optimized this way?
I've changes to GPLv3, because in root directory there is COPYING file which states so. I wanted just to disambiguate it in files. Of course I can change it to GPLv2.

**Comment by andreasb242** [comment]
@wbrenna You copied the License text of the GPL3 to the COPYING file, I only wrote GPL in the Source Code, did you choose this version for a special case or only because it was the newest version?

As I read, the GPLv3 has a lot of issues, which the Version 2 does not have. Would it be Ok to change it to GPLv2 or Later?

@MarPiRK I created you the string_new, you can push, then I'll have a look at the Code, OK?
I hope it's OK for wbrenna, I did 3 Years no development on Xournal++.

After my study I had really enough from note taking Software ;-)

**Comment by MarPiRK** [comment]
Regarding http://stackoverflow.com/a/18892532 benchmark it seems, that stringstreams are very unefficient in concatenating strings. I don't know which place in code you really mean, so please point an example of changing string in loop.

**Comment by andreasb242** [comment]
All Ok, you're right.

**Comment by wbrenna** [comment]
I picked v3 because just GPL was ambiguous so I had to choose one. I do like some of the provisions of v3 but I understand and agree that there is some uncertainty about key sharing in particular, so v2 or later is fine by me. I would prefer the "or later" clause, so that it can be upgraded in the future (for example, if v4 clarifies key sharing).

Glad to have you back!

**Comment by MarPiRK** [comment]
I think that it can be closed.

**Comment by corna** [comment]
Yes, now everything works perfectly
Thanks for the help
